File: android/app/src/main/java/dev/mazgi/app/AuthViewModel.kt
package dev.mazgi.app

import android.app.Application
import android.content.Context
import android.net.Uri
import androidx.browser.customtabs.CustomTabsIntent
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import org.json.JSONObject

enum class MfaStep { IDLE, SETUP, RECOVERY, DISABLE, REGENERATE }

enum class ThemeMode(val labelResId: Int) {
    SYSTEM(R.string.theme_system),
    LIGHT(R.string.theme_light),
    DARK(R.string.theme_dark);

    val apiValue: String get() = name.lowercase()

    companion object {
        fun fromApi(value: String?): ThemeMode = when (value) {
            "light" -> LIGHT
            "dark" -> DARK
            else -> SYSTEM
        }
    }
}

data class AuthUiState(
    val user: UserProfile? = null,
    val accessToken: String? = null,
    val isAuthenticated: Boolean = false,
    val isLoading: Boolean = false,
    val isRestoringSession: Boolean = true,
    val errorMessage: String? = null,
    val themeMode: ThemeMode = ThemeMode.SYSTEM,
    val verificationSentEmail: String? = null,
    val mfaToken: String? = null,
    val totpSetupUri: String? = null,
    val totpSetupSecret: String? = null,
    val recoveryCodes: List<String>? = null,
    val mfaStep: MfaStep = MfaStep.IDLE,
    val passwordResetSent: Boolean = false,
)

class AuthViewModel(application: Application) : AndroidViewModel(application) {

    private val _uiState = MutableStateFlow(AuthUiState())
    val uiState: StateFlow<AuthUiState> = _uiState.asStateFlow()

    private val api = APIClient.shared
    private val prefs = application.getSharedPreferences("auth", Context.MODE_PRIVATE)

    init {
        viewModelScope.launch {
            restoreSession()
            _uiState.value = _uiState.value.copy(isRestoringSession = false)
        }
    }

    fun signIn(email: String, password: String) {
        viewModelScope.launch {
            perform {
                when (val result = api.signIn(email, password)) {
                    is SignInResult.Success -> {
                        val res = result.response
                        storeTokens(res.accessToken, res.refreshToken)
                        _uiState.value = _uiState.value.copy(user = res.user, accessToken = res.accessToken, isAuthenticated = true)
                        syncTheme(res.user)
                    }
                    is SignInResult.MfaRequired -> {
                        _uiState.value = _uiState.value.copy(mfaToken = result.mfaToken)
                    }
                }
            }
        }
    }

    fun signUp(email: String, password: String) {
        viewModelScope.launch {
            perform {
                api.signUp(email, password)
                _uiState.value = _uiState.value.copy(verificationSentEmail = email)
            }
        }
    }

    fun resendVerification(email: String) {
        viewModelScope.launch {
            perform {
                api.resendVerification(email)
            }
        }
    }

    fun clearVerificationSent() {
        _uiState.value = _uiState.value.copy(verificationSentEmail = null)
    }

    fun signInWithApple(context: Context) {
        val uri = Uri.parse("${api.baseUrl}/auth/apple/native")
        try {
            CustomTabsIntent.Builder().build().launchUrl(context, uri)
        } catch (e: Exception) {
            _uiState.value = _uiState.value.copy(errorMessage = e.message ?: "Failed to open browser")
        }
    }

    fun signInWithGithub(context: Context) {
        val uri = Uri.parse("${api.baseUrl}/auth/github/native")
        try {
            CustomTabsIntent.Builder().build().launchUrl(context, uri)
        } catch (e: Exception) {
            _uiState.value = _uiState.value.copy(errorMessage = e.message ?: "Failed to open browser")
        }
    }

    fun signInWithGoogle(context: Context) {
        val uri = Uri.parse("${api.baseUrl}/auth/google/native")
        try {
            CustomTabsIntent.Builder().build().launchUrl(context, uri)
        } catch (e: Exception) {
            _uiState.value = _uiState.value.copy(errorMessage = e.message ?: "Failed to open browser")
        }
    }

    fun signInWithTwitter(context: Context) {
        // Use TWITTER_AUTH_BASE_URL (defaults to localhost via `adb reverse tcp:4000 tcp:4000`)
        // so the session cookie domain matches the TWITTER_NATIVE_CALLBACK_URL.
        // If 10.0.2.2 were used here, the PKCE code_verifier stored in express-session would be
        // lost at the callback step because the browser won't send a 10.0.2.2 cookie to localhost.
        val uri = Uri.parse("${api.twitterAuthBaseUrl}/auth/twitter/native")
        try {
            CustomTabsIntent.Builder().build().launchUrl(context, uri)
        } catch (e: Exception) {
            _uiState.value = _uiState.value.copy(errorMessage = e.message ?: "Failed to open browser")
        }
    }

    fun signInWithDiscord(context: Context) {
        val uri = Uri.parse("${api.baseUrl}/auth/discord/native")
        try {
            CustomTabsIntent.Builder().build().launchUrl(context, uri)
        } catch (e: Exception) {
            _uiState.value = _uiState.value.copy(errorMessage = e.message ?: "Failed to open browser")
        }
    }

    fun linkProvider(context: Context, provider: String) {
        val token = _uiState.value.accessToken ?: return
        // For Twitter, must use TWITTER_AUTH_BASE_URL (session cookie domain consistency for PKCE)
        val baseUrl = if (provider == "twitter") api.twitterAuthBaseUrl else api.baseUrl
        val uri = Uri.parse("$baseUrl/auth/link/$provider/native?token=$token")
        try {
            CustomTabsIntent.Builder().build().launchUrl(context, uri)
        } catch (e: Exception) {
            _uiState.value = _uiState.value.copy(errorMessage = e.message ?: "Failed to open browser")
        }
    }

    fun unlinkProvider(provider: String) {
        viewModelScope.launch {
            val token = _uiState.value.accessToken ?: return@launch
            _uiState.value = _uiState.value.copy(isLoading = true, errorMessage = null)
            try {
                val updatedUser = api.unlinkProvider(token, provider)
                _uiState.value = _uiState.value.copy(user = updatedUser, isLoading = false)
            } catch (e: APIException) {
                _uiState.value = _uiState.value.copy(errorMessage = e.message, isLoading = false)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(errorMessage = e.message ?: "Unknown error", isLoading = false)
            }
        }
    }

    fun updateEmail(newEmail: String) {
        viewModelScope.launch {
            val token = _uiState.value.accessToken ?: return@launch
            _uiState.value = _uiState.value.copy(isLoading = true, errorMessage = null)
            try {
                val updatedUser = api.updateEmail(token, newEmail)
                _uiState.value = _uiState.value.copy(user = updatedUser, isLoading = false)
            } catch (e: APIException) {
                _uiState.value = _uiState.value.copy(errorMessage = e.message, isLoading = false)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(errorMessage = e.message ?: "Unknown error", isLoading = false)
            }
        }
    }

    fun resendVerificationFromSettings() {
        viewModelScope.launch {
            val email = _uiState.value.user?.email ?: return@launch
            _uiState.value = _uiState.value.copy(isLoading = true, errorMessage = null)
            try {
                api.resendVerification(email)
                _uiState.value = _uiState.value.copy(isLoading = false)
            } catch (e: APIException) {
                _uiState.value = _uiState.value.copy(errorMessage = e.message, isLoading = false)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(errorMessage = e.message ?: "Unknown error", isLoading = false)
            }
        }
    }

    fun forgotPasswordFromSignIn(email: String) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, errorMessage = null)
            try {
                api.forgotPassword(email)
                _uiState.value = _uiState.value.copy(passwordResetSent = true, isLoading = false)
            } catch (e: APIException) {
                // Always show success to prevent email enumeration
                _uiState.value = _uiState.value.copy(passwordResetSent = true, isLoading = false)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(passwordResetSent = true, isLoading = false)
            }
        }
    }

    fun requestPasswordReset() {
        viewModelScope.launch {
            val email = _uiState.value.user?.email ?: return@launch
            _uiState.value = _uiState.value.copy(isLoading = true, errorMessage = null)
            try {
                api.forgotPassword(email)
                _uiState.value = _uiState.value.copy(isLoading = false)
            } catch (e: APIException) {
                _uiState.value = _uiState.value.copy(errorMessage = e.message, isLoading = false)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(errorMessage = e.message ?: "Unknown error", isLoading = false)
            }
        }
    }

    fun deleteAccount() {
        viewModelScope.launch {
            val token = _uiState.value.accessToken ?: return@launch
            _uiState.value = _uiState.value.copy(isLoading = true, errorMessage = null)
            try {
                api.deleteAccount(token)
                signOut()
            } catch (e: APIException) {
                _uiState.value = _uiState.value.copy(errorMessage = e.message, isLoading = false)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(errorMessage = e.message ?: "Unknown error", isLoading = false)
            }
        }
    }

    fun handleOAuthCallback(uri: Uri) {
        // Account linking callback — refresh user profile
        val linked = uri.getQueryParameter("linked")
        if (linked != null) {
            viewModelScope.launch {
                val token = _uiState.value.accessToken ?: return@launch
                try {
                    val profile = api.me(token)
                    _uiState.value = _uiState.value.copy(user = profile)
                } catch (_: Exception) { }
            }
            return
        }

        val accessToken = uri.getQueryParameter("accessToken") ?: return
        val refreshToken = uri.getQueryParameter("refreshToken") ?: return
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, errorMessage = null)
            try {
                val profile = api.me(accessToken)
                storeTokens(accessToken, refreshToken)
                _uiState.value = _uiState.value.copy(
                    user = profile,
                    accessToken = accessToken,
                    isAuthenticated = true,
                    isLoading = false,
                )
                syncTheme(profile)
            } catch (e: APIException) {
                _uiState.value = _uiState.value.copy(errorMessage = e.message, isLoading = false)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    errorMessage = e.message ?: "Unknown error",
                    isLoading = false,
                )
            }
        }
    }

    fun setTheme(mode: ThemeMode) {
        _uiState.value = _uiState.value.copy(themeMode = mode)
        viewModelScope.launch {
            val token = _uiState.value.accessToken ?: return@launch
            try {
                val body = JSONObject().apply { put("theme", mode.apiValue) }
                val updated = api.updatePreferences(token, body)
                _uiState.value = _uiState.value.copy(user = updated)
            } catch (_: Exception) {
                // Theme already applied locally; ignore network errors
            }
        }
    }

    fun verifyMfa(code: String) {
        viewModelScope.launch {
            val token = _uiState.value.mfaToken ?: return@launch
            perform {
                val res = api.totpVerify(token, code)
                storeTokens(res.accessToken, res.refreshToken)
                _uiState.value = _uiState.value.copy(
                    user = res.user,
                    accessToken = res.accessToken,
                    isAuthenticated = true,
                    mfaToken = null,
                )
                syncTheme(res.user)
            }
        }
    }

    fun setupTotp() {
        viewModelScope.launch {
            val token = _uiState.value.accessToken ?: return@launch
            _uiState.value = _uiState.value.copy(isLoading = true, errorMessage = null)
            try {
                val setup = api.totpSetup(token)
                _uiState.value = _uiState.value.copy(
                    totpSetupUri = setup.uri,
                    totpSetupSecret = setup.secret,
                    mfaStep = MfaStep.SETUP,
                    isLoading = false,
                )
            } catch (e: APIException) {
                _uiState.value = _uiState.value.copy(errorMessage = e.message, isLoading = false)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(errorMessage = e.message ?: "Unknown error", isLoading = false)
            }
        }
    }

    fun enableTotp(code: String) {
        viewModelScope.launch {
            val token = _uiState.value.accessToken ?: return@launch
            _uiState.value = _uiState.value.copy(isLoading = true, errorMessage = null)
            try {
                val result = api.totpEnable(token, code)
                val profile = api.me(token)
                _uiState.value = _uiState.value.copy(
                    user = profile,
                    recoveryCodes = result.recoveryCodes,
                    mfaStep = MfaStep.RECOVERY,
                    totpSetupUri = null,
                    totpSetupSecret = null,
                    isLoading = false,
                )
            } catch (e: APIException) {
                _uiState.value = _uiState.value.copy(errorMessage = e.message, isLoading = false)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(errorMessage = e.message ?: "Unknown error", isLoading = false)
            }
        }
    }

    fun disableTotp(code: String) {
        viewModelScope.launch {
            val token = _uiState.value.accessToken ?: return@launch
            _uiState.value = _uiState.value.copy(isLoading = true, errorMessage = null)
            try {
                api.totpDisable(token, code)
                val profile = api.me(token)
                _uiState.value = _uiState.value.copy(
                    user = profile,
                    mfaStep = MfaStep.IDLE,
                    isLoading = false,
                )
            } catch (e: APIException) {
                _uiState.value = _uiState.value.copy(errorMessage = e.message, isLoading = false)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(errorMessage = e.message ?: "Unknown error", isLoading = false)
            }
        }
    }

    fun regenerateRecoveryCodes(code: String) {
        viewModelScope.launch {
            val token = _uiState.value.accessToken ?: return@launch
            _uiState.value = _uiState.value.copy(isLoading = true, errorMessage = null)
            try {
                val result = api.totpRegenerateRecoveryCodes(token, code)
                _uiState.value = _uiState.value.copy(
                    recoveryCodes = result.recoveryCodes,
                    mfaStep = MfaStep.RECOVERY,
                    isLoading = false,
                )
            } catch (e: APIException) {
                _uiState.value = _uiState.value.copy(errorMessage = e.message, isLoading = false)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(errorMessage = e.message ?: "Unknown error", isLoading = false)
            }
        }
    }

    fun startDisableTotp() {
        _uiState.value = _uiState.value.copy(mfaStep = MfaStep.DISABLE, errorMessage = null)
    }

    fun startRegenerateRecoveryCodes() {
        _uiState.value = _uiState.value.copy(mfaStep = MfaStep.REGENERATE, errorMessage = null)
    }

    fun clearMfaStep() {
        _uiState.value = _uiState.value.copy(
            mfaStep = MfaStep.IDLE,
            totpSetupUri = null,
            totpSetupSecret = null,
            recoveryCodes = null,
        )
    }

    fun signOut() {
        prefs.edit().remove("access_token").remove("refresh_token").apply()
        _uiState.value = AuthUiState(isRestoringSession = false, passwordResetSent = false)
    }

    fun clearError() {
        _uiState.value = _uiState.value.copy(errorMessage = null)
    }

    // MARK: - Session restore

    private suspend fun restoreSession() {
        val accessToken = prefs.getString("access_token", null) ?: return
        val refreshToken = prefs.getString("refresh_token", null) ?: return
        try {
            val profile = api.me(accessToken)
            _uiState.value = _uiState.value.copy(user = profile, accessToken = accessToken, isAuthenticated = true)
            syncTheme(profile)
        } catch (e: Exception) {
            // Access token expired — try refresh
            try {
                val res = api.refresh(refreshToken)
                storeTokens(res.accessToken, res.refreshToken)
                _uiState.value = _uiState.value.copy(user = res.user, accessToken = res.accessToken, isAuthenticated = true)
                syncTheme(res.user)
            } catch (e2: Exception) {
                signOut()
            }
        }
    }

    // MARK: - Helpers

    private suspend fun perform(action: suspend () -> Unit) {
        _uiState.value = _uiState.value.copy(isLoading = true, errorMessage = null)
        try {
            action()
        } catch (e: APIException) {
            _uiState.value = _uiState.value.copy(errorMessage = e.message)
        } catch (e: Exception) {
            _uiState.value = _uiState.value.copy(errorMessage = e.message ?: "Unknown error")
        } finally {
            _uiState.value = _uiState.value.copy(isLoading = false)
        }
    }

    private fun storeTokens(accessToken: String, refreshToken: String) {
        prefs.edit()
            .putString("access_token", accessToken)
            .putString("refresh_token", refreshToken)
            .apply()
    }

    private fun syncTheme(profile: UserProfile) {
        val mode = ThemeMode.fromApi(profile.preferences?.theme)
        _uiState.value = _uiState.value.copy(themeMode = mode)
    }
}

File: apple/app/AuthViewModel.swift
import Foundation
import Observation
import SwiftUI

enum ThemeMode: String, CaseIterable {
    case system, light, dark

    var colorScheme: ColorScheme? {
        switch self {
        case .system: return nil
        case .light: return .light
        case .dark: return .dark
        }
    }

    var label: String {
        switch self {
        case .system: return "System"
        case .light: return "Light"
        case .dark: return "Dark"
        }
    }
}

enum MfaStep {
    case idle, setup, recovery, disable, regenerate
}

@Observable
@MainActor
final class AuthViewModel {
    private(set) var user: UserProfile?
    private(set) var isAuthenticated = false
    private(set) var accessToken: String?
    private(set) var themeMode: ThemeMode = .system
    var isLoading = false
    var errorMessage: String?
    private(set) var verificationSentEmail: String?
    private(set) var mfaToken: String?
    private(set) var totpSetupUri: String?
    private(set) var totpSetupSecret: String?
    private(set) var recoveryCodes: [String]?
    private(set) var mfaStep: MfaStep = .idle
    private(set) var passwordResetSent = false

    private let api: APIClient
    private let tokensKey = "auth_tokens"

    private struct StoredTokens: Codable {
        let accessToken: String
        let refreshToken: String
    }

    init(api: APIClient = .shared) {
        self.api = api
        Task { await restoreSession() }
    }

    // MARK: - Public actions

    func signIn(email: String, password: String) async {
        await perform {
            let result = try await self.api.signIn(email: email, password: password)
            switch result {
            case .success(let res):
                self.mfaToken = nil
                self.store(res)
            case .mfaRequired(let token):
                self.mfaToken = token
            }
        }
    }

    func signUp(email: String, password: String) async {
        await perform {
            _ = try await self.api.signUp(email: email, password: password)
            self.verificationSentEmail = email
        }
    }

    func resendVerification(email: String) async {
        await perform {
            _ = try await self.api.resendVerification(email: email)
        }
    }

    func clearVerificationSent() {
        verificationSentEmail = nil
    }

    func signInWithTwitter() async {
        await signInWithOAuth(provider: "twitter")
    }

    func signInWithGithub() async {
        await signInWithOAuth(provider: "github")
    }

    func signInWithGoogle() async {
        await signInWithOAuth(provider: "google")
    }

    func signInWithApple() async {
        await signInWithOAuth(provider: "apple")
    }

    func signInWithDiscord() async {
        await signInWithOAuth(provider: "discord")
    }

    private func signInWithOAuth(provider: String) async {
        guard let url = URL(string: "\(api.baseURL)/auth/\(provider)/native") else { return }
        isLoading = true
        errorMessage = nil
        defer { isLoading = false }
        do {
            let webAuth = WebAuthSession()
            guard let callbackURL = try await webAuth.authenticate(
                url: url,
                callbackURLScheme: "oauth2app"
            ) else {
                return // User cancelled
            }
            let components = URLComponents(url: callbackURL, resolvingAgainstBaseURL: false)
            let items = components?.queryItems ?? []

            guard
                let accessToken = items.first(where: { $0.name == "accessToken" })?.value,
                let refreshToken = items.first(where: { $0.name == "refreshToken" })?.value
            else {
                throw APIError(statusCode: 0, message: "Invalid OAuth callback URL")
            }
            let profile = try await api.me(accessToken: accessToken)
            store(AuthResponse(accessToken: accessToken, refreshToken: refreshToken, user: profile))
        } catch let e as APIError {
            errorMessage = e.message
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func linkProvider(_ provider: String) async {
        guard let token = accessToken,
              let url = URL(string: "\(api.baseURL)/auth/link/\(provider)/native?token=\(token)") else { return }
        isLoading = true
        errorMessage = nil
        defer { isLoading = false }
        do {
            let webAuth = WebAuthSession()
            guard let callbackURL = try await webAuth.authenticate(
                url: url,
                callbackURLScheme: "oauth2app"
            ) else {
                return // User cancelled
            }
            // Check if this was a link callback (has ?linked= param)
            let components = URLComponents(url: callbackURL, resolvingAgainstBaseURL: false)
            let items = components?.queryItems ?? []
            if items.first(where: { $0.name == "linked" }) != nil {
                // Refresh user profile to get updated linked accounts
                if let profile = try? await api.me(accessToken: token) {
                    user = profile
                }
            }
        } catch let e as APIError {
            errorMessage = e.message
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func unlinkProvider(_ provider: String) async {
        guard let token = accessToken else { return }
        isLoading = true
        errorMessage = nil
        defer { isLoading = false }
        do {
            let updatedUser = try await api.unlinkProvider(accessToken: token, provider: provider)
            user = updatedUser
        } catch let e as APIError {
            errorMessage = e.message
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func updateEmail(_ newEmail: String) async {
        guard let token = accessToken else { return }
        isLoading = true
        errorMessage = nil
        defer { isLoading = false }
        do {
            let updatedUser = try await api.updateEmail(accessToken: token, email: newEmail)
            user = updatedUser
        } catch let e as APIError {
            errorMessage = e.message
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func resendVerificationFromSettings() async {
        guard let email = user?.email else { return }
        await perform {
            _ = try await self.api.resendVerification(email: email)
        }
    }

    func requestPasswordReset() async {
        guard let email = user?.email else { return }
        await perform {
            _ = try await self.api.forgotPassword(email: email)
        }
    }

    func requestPasswordResetFromSignIn(email: String) async {
        await perform {
            _ = try await self.api.forgotPassword(email: email)
            self.passwordResetSent = true
        }
    }

    func deleteAccount() async {
        guard let token = accessToken else { return }
        isLoading = true
        errorMessage = nil
        defer { isLoading = false }
        do {
            try await api.deleteAccount(accessToken: token)
            signOut()
        } catch let e as APIError {
            errorMessage = e.message
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func setTheme(_ mode: ThemeMode) async {
        themeMode = mode
        guard let token = accessToken else { return }
        do {
            let updated = try await api.updatePreferences(
                accessToken: token,
                preferences: UserPreferences(theme: mode.rawValue)
            )
            user = updated
        } catch {
            // Theme already applied locally; ignore network errors
        }
    }

    // MARK: - MFA

    func verifyMfa(code: String) async {
        guard let token = mfaToken else { return }
        await perform {
            let res = try await self.api.totpVerify(mfaToken: token, code: code)
            self.mfaToken = nil
            self.store(res)
        }
    }

    func setupTotp() async {
        guard let token = accessToken else { return }
        await perform {
            let res = try await self.api.totpSetup(accessToken: token)
            self.totpSetupUri = res.uri
            self.totpSetupSecret = res.secret
            self.mfaStep = .setup
        }
    }

    func enableTotp(code: String) async {
        guard let token = accessToken else { return }
        await perform {
            let res = try await self.api.totpEnable(accessToken: token, code: code)
            self.recoveryCodes = res.recoveryCodes
            self.totpSetupUri = nil
            self.totpSetupSecret = nil
            self.mfaStep = .recovery
            // Refresh user profile to reflect totpEnabled
            if let profile = try? await self.api.me(accessToken: token) {
                self.user = profile
            }
        }
    }

    func disableTotp(code: String) async {
        guard let token = accessToken else { return }
        await perform {
            _ = try await self.api.totpDisable(accessToken: token, code: code)
            self.mfaStep = .idle
            // Refresh user profile to reflect totpEnabled
            if let profile = try? await self.api.me(accessToken: token) {
                self.user = profile
            }
        }
    }

    func regenerateRecoveryCodes(code: String) async {
        guard let token = accessToken else { return }
        await perform {
            let res = try await self.api.totpRegenerateRecoveryCodes(accessToken: token, code: code)
            self.recoveryCodes = res.recoveryCodes
            self.mfaStep = .recovery
        }
    }

    func clearMfaStep() {
        mfaStep = .idle
        totpSetupUri = nil
        totpSetupSecret = nil
        recoveryCodes = nil
    }

    func beginDisableTotp() {
        mfaStep = .disable
    }

    func beginRegenerateRecoveryCodes() {
        mfaStep = .regenerate
    }

    func signOut() {
        UserDefaults.standard.removeObject(forKey: tokensKey)
        user = nil
        accessToken = nil
        isAuthenticated = false
        themeMode = .system
        mfaToken = nil
        mfaStep = .idle
        totpSetupUri = nil
        totpSetupSecret = nil
        recoveryCodes = nil
        passwordResetSent = false
    }

    // MARK: - Session restore

    private func restoreSession() async {
        guard let tokens = loadTokens() else { return }
        do {
            let profile = try await api.me(accessToken: tokens.accessToken)
            user = profile
            accessToken = tokens.accessToken
            isAuthenticated = true
            syncTheme(from: profile)
        } catch {
            // Access token expired — try refresh
            do {
                let res = try await api.refresh(refreshToken: tokens.refreshToken)
                store(res)
            } catch {
                signOut()
            }
        }
    }

    // MARK: - Helpers

    private func perform(_ action: () async throws -> Void) async {
        isLoading = true
        errorMessage = nil
        defer { isLoading = false }
        do {
            try await action()
        } catch let e as APIError {
            errorMessage = e.message
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    private func store(_ res: AuthResponse) {
        let tokens = StoredTokens(accessToken: res.accessToken, refreshToken: res.refreshToken)
        if let data = try? JSONEncoder().encode(tokens) {
            UserDefaults.standard.set(data, forKey: tokensKey)
        }
        user = res.user
        accessToken = res.accessToken
        isAuthenticated = true
        syncTheme(from: res.user)
    }

    private func syncTheme(from profile: UserProfile) {
        if let raw = profile.preferences?.theme, let mode = ThemeMode(rawValue: raw) {
            themeMode = mode
        } else {
            themeMode = .system
        }
    }

    private func loadTokens() -> StoredTokens? {
        guard let data = UserDefaults.standard.data(forKey: tokensKey) else { return nil }
        return try? JSONDecoder().decode(StoredTokens.self, from: data)
    }
}
